### PR TITLE
Perform case-sensitive search for files to prevent accidental .cmi shadowing

### PR DESCRIPTION
**Pull request by dra27**
When a file-system is set to be case-insensitive, it is possible for a .cmi file to shadow another .cmi file further down the search path (for example, a `uChar.cmi` can shadow the standard library's `uchar.cmi`). This is a long-standing problem on both Windows and macOS.

This changes the search to be based on `Sys.readdir` instead of `Sys.file_exists` with a cache to prevent too many calls to the function.

An escape switch `Misc.reset_find_cache` allows the toplevel to clear this cache before each phrase is parsed, meaning that a long-lived toploop can still see changes made externally.

I've taken the view that any use of those Misc.find_ functions can be safely switched to be case-sensitive, given that in so many instances it's necessary for the case of the filename to be correct for it to be used anyway.

I've gone through places in the codebase where `Sys.file_exists` is used directly and identified a few where `Misc.really_exists` should probably be used instead, but I'd like the implementation as it stands now to be reviewed before extending it to this.

This GPR means that Unix and Windows/macOS now behave the same way when presented with `FOo.ml` containing `let answer = 42` and `Bar.ml` containing `print_int Foo.answer` when running `ocamlc -c FOo.ml Bar.ml` in that both now display:

```
File "Bar.ml", line 1, characters 29-39:
Error: Unbound module Foo
```
where before Windows and macOS would display:
```
File "Bar.ml", line 1:
Error: Wrong file naming: foo.cmi contains the compiled interface for
FOo when Foo was expected
```

### CODE REVIEW

**Comment by gasche** [comment]
Any numbers on the performance impact of the change?

(Edit: I closed the PR by mistake, sorry.)

**Comment by dra27** [comment]
Not yet, no - just that anecdotally my Windows build doesn't seem to be any slower than it was before!

**Comment by dra27** [comment]
(However, I will do some proper testing of that later!)

**Comment by gasche** [comment]
> An escape switch Misc.reset_find_cache allows the toplevel to clear this cache before each phrase is parsed, meaning that a long-lived toploop can still see changes made externally.

The Merlin people (cc @let-def and @trefis) have a cache for loaded `.cmi`s as well and an invalidation strategy, maybe they will be interested in hearing about this new caching layer.

File: utils/misc.ml [hunk 1]
```diff
@@ -181,15 +181,50 @@ let may_map = Stdlib.Option.map
...
+  Hashtbl.clear find_cache
+
+let really_exists dir name =
+  let dir = if dir = "" then Filename.current_dir_name else dir in
+  let () =
+    if Filename.is_relative dir then
+      let cwd = Sys.getcwd () in
+      if !find_cache_cwd <> cwd then begin
+        Hashtbl.clear find_cache;
+        find_cache_cwd := cwd
```
Comment (ghost) [inline utils/misc.ml:196]: Why not use `let dir = Filename.concat cwd dir in` in that case rather than clear the cache?

File: utils/misc.ml [hunk 2]
```diff
@@ -181,15 +181,50 @@ let may_map = Stdlib.Option.map
...
+  | files -> Hashtbl.mem files name
+  | exception Not_found ->
+      try
+        let files = Sys.readdir dir in
+        let table = Hashtbl.create (Array.length files * 2) in
+        let add found file =
+          Hashtbl.add table file (); found || file = name
+        in
+        Hashtbl.add find_cache dir table;
+        Array.fold_left add false files
```
Comment (ghost) [inline utils/misc.ml:209]: It seems simpler to me to just use a string set here

File: utils/misc.ml [hunk 3]
```diff
@@ -181,15 +181,50 @@ let may_map = Stdlib.Option.map
...
+    else
+      raise Not_found
   else begin
     let rec try_dir = function
       [] -> raise Not_found
     | dir::rem ->
-        let fullname = Filename.concat dir name in
-        if Sys.file_exists fullname then fullname else try_dir rem
+        let name = Filename.concat dir name in
+        if really_exists (Filename.dirname name) (Filename.basename name) then
```
Comment (ghost) [inline utils/misc.ml:224]: Sometimes you do `if really_exists (Filename.dirname name) (Filename.basename name) then name` and sometimes `if really_exists dir name then Filename.concat dir name`. I find it a bit hard to follow, why not just pass a single filename to `really_exists`?

**Comment by dra27** [comment]
The primary motivation for this PR is subsumed by #2041, however a quick check suggests that there are some uses of `Sys.file_exists` remaining in the codebase which may need checking further, so I'm leaving this open for now.

**Comment by gasche** [comment]
@dra27 has apparently been unable to come back to this, and has many other interesting and important things on his plate, so let's close for now. We can always reopen for the next load-path-case-sensitivity revival.
